packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts

```diff
@@ -0,0 +1,196 @@
+  return new Promise(resolve => {
+    cordova.plugins.browsertab.isAvailable(browserTabIsAvailable => {
+      if (browserTabIsAvailable) {
+        cordova.plugins.browsertab.openUrl(handlerUrl);
+      } else {
+        // TODO: Return the inappbrowser ref that's returned from the open call
+        cordova.InAppBrowser.open(
+          handlerUrl,
+          _isIOS7Or8() ? '_blank' : '_system',
+          'location=yes'
+        );
+      }
+      resolve();
+    });
+  });
+}
+
+/**
+ * Checks the configuration of the Cordova environment. This has no side effect
+ * if the configuration is correct; otherwise it throws an error with the
+ * missing plugin.
+ */
+export function _checkCordovaConfiguration(auth: Auth): void {
+  // Check all dependencies installed.
+  // https://github.com/nordnet/cordova-universal-links-plugin
+  // Note that cordova-universal-links-plugin has been abandoned.
+  // A fork with latest fixes is available at:
+  // https://www.npmjs.com/package/cordova-universal-links-plugin-fix
+  _assert(
+    typeof window?.universalLinks?.subscribe === 'function',
+    auth,
+    AuthErrorCode.INVALID_CORDOVA_CONFIGURATION,
+    {
+      missingPlugin: 'cordova-universal-links-plugin-fix'
+    }
+  );
+
+  // https://www.npmjs.com/package/cordova-plugin-buildinfo
+  _assert(
+    typeof window?.BuildInfo?.packageName !== 'undefined',
+    auth,
+    AuthErrorCode.INVALID_CORDOVA_CONFIGURATION,
+    {
+      missingPlugin: 'cordova-plugin-buildInfo'
+    }
+  );
+
+  // https://github.com/google/cordova-plugin-browsertab
+  _assert(
+    typeof window?.cordova?.plugins?.browsertab?.openUrl === 'function',
+    auth,
+    AuthErrorCode.INVALID_CORDOVA_CONFIGURATION,
+    {
+      missingPlugin: 'cordova-plugin-browsertab'
+    }
+  );
+  _assert(
+    typeof window?.cordova?.plugins?.browsertab?.isAvailable === 'function',
+    auth,
+    AuthErrorCode.INVALID_CORDOVA_CONFIGURATION,
+    {
+      missingPlugin: 'cordova-plugin-browsertab'
+    }
+  );
+
+  // https://cordova.apache.org/docs/en/latest/reference/cordova-plugin-inappbrowser/
+  _assert(
+    typeof window?.cordova?.InAppBrowser?.open === 'function',
+    auth,
+    AuthErrorCode.INVALID_CORDOVA_CONFIGURATION,
+    {
+      missingPlugin: 'cordova-plugin-inappbrowser'
+    }
+  );
+}
+
+function generateSessionId(): string {
+  const chars = [];
+  const allowedChars =
+    '1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
+  for (let i = 0; i < SESSION_ID_LENGTH; i++) {
+    const idx = Math.floor(Math.random() * allowedChars.length);
+    chars.push(allowedChars.charAt(idx));
+  }
+  return chars.join('');
+}
+
+/**
+ * Computes the SHA-256 of a session ID. The SubtleCrypto interface is only
+ * available in "secure" contexts, which covers Cordova (which is served on a file
+ * protocol).
+ */
+async function computeSha256(sessionId: string): Promise<string> {
+  const encoder = new TextEncoder();
```

Comment:
What do we do for IE11? https://caniuse.com/textencoder says no support

Reply:
I didn't realize Cordova supported IE11 :(

Reply:
Added a function to fallback to a simple encoding function